src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs

```diff
@@ -0,0 +1,582 @@
+		NSObject? curveSize = null;
+
+		var foundAnimationDuration = notification.UserInfo?.TryGetValue(new NSString("UIKeyboardAnimationDurationUserInfoKey"), out curveSize);
+		if (foundAnimationDuration == true && curveSize is not null)
+		{
+			var curveNum = (NSNumber)NSObject.FromObject(curveSize);
+			var num = (double)curveNum;
+			if (num != 0)
+				AnimationDuration = num;
+		}
+
+		if (LastScrollView is not null)
+			UIView.Animate(AnimationDuration, 0, UIViewAnimationOptions.CurveEaseOut, AnimateHidingKeyboard, () => { });
+
+		if (IsKeyboardShowing)
+			RestorePosition();
+
+		IsKeyboardShowing = false;
+		View = null;
+		LastScrollView = null;
+		KeyboardFrame = CGRect.Empty;
+		StartingContentInsets = new UIEdgeInsets();
+		StartingScrollIndicatorInsets = new UIEdgeInsets();
+		StartingContentInsets = new UIEdgeInsets();
+	}
+
+	static void AnimateHidingKeyboard()
+	{
+		if (LastScrollView is not null && LastScrollView.ContentInset != StartingContentInsets)
+		{
+			LastScrollView.ContentInset = StartingContentInsets;
+			LastScrollView.ScrollIndicatorInsets = StartingScrollIndicatorInsets;
+		}
+
+		var superScrollView = LastScrollView;
+		while (superScrollView is not null)
+		{
+			var contentSize = new CGSize(Math.Max(superScrollView.ContentSize.Width, superScrollView.Frame.Width),
+				Math.Max(superScrollView.ContentSize.Height, superScrollView.Frame.Height));
+
+			var minY = contentSize.Height - superScrollView.Frame.Height;
+			if (minY < superScrollView.ContentOffset.Y)
+			{
+				var newContentOffset = new CGPoint(superScrollView.ContentOffset.X, minY);
+				if (!superScrollView.ContentOffset.Equals(newContentOffset))
+				{
+					if (View?.Superview is UIStackView)
+						superScrollView.SetContentOffset(newContentOffset, UIView.AnimationsEnabled);
+					else
+						superScrollView.ContentOffset = newContentOffset;
+				}
+			}
+			superScrollView = superScrollView.FindResponder<UIScrollView>();
+		}
+	}
+
+	internal static void Destroy()
+	{
+		if (WillShowToken is not null)
+			NSNotificationCenter.DefaultCenter.RemoveObserver(WillShowToken);
+		if (WillHideToken is not null)
+			NSNotificationCenter.DefaultCenter.RemoveObserver(WillHideToken);
+		if (TextFieldToken is not null)
+			NSNotificationCenter.DefaultCenter.RemoveObserver(TextFieldToken);
+		if (TextViewToken is not null)
+			NSNotificationCenter.DefaultCenter.RemoveObserver(TextViewToken);
+	}
+
+	// Used to get the numeric values from the UserInfo dictionary's NSObject value to CGRect.
+	// Doing manually since CGRectFromString is not yet bound
+	static CGRect? DescriptionToCGRect(string? description)
+	{
+		// example of passed in description: "NSRect: {{0, 586}, {430, 346}}"
+
+		if (description is null)
+			return null;
+
+		// remove everything except for numbers and commas
+		var temp = Regex.Replace(description, @"[^0-9,]", "");
+		var dimensions = temp.Split(',');
+
+		if (int.TryParse(dimensions[0], out var x) && int.TryParse(dimensions[1], out var y)
```

Comment:
Lets be ore defensive and check the size of the dimensions before we try to get the values so that we do not have an index out of bounds. If the size is not correct, return null

Reply:
I'm not sure about this one. It looks like these should actually be CGFloat -> nfloat's and the apple documentation says they can be negative as well. Other than changing these from ints to nfloats, are there other things that can make this out of bounds?

Reply:
I mean, dimensions, the array from the regexp is not guaranteed to have 1 or more matches (should, but there we are not 100% sure). So I would simply add a:
```csharp
if (dimensions.Lenth != 2) {
    return null;
}
```

Reply:
AH Sorry! Will do!